contrib/python/src/python/pants/contrib/python/checks/tasks/checkstyle/checker.py

```diff
@@ -175,17 +173,42 @@ def execute(self):
     # If we are linting for Python 3, skip lint altogether.
     # Long-term Python 3 linting solution tracked by:
     # https://github.com/pantsbuild/pants/issues/5764
-    interpreter = self.context.products.get_data(PythonInterpreter)
-    # Check interpreter is not 'None' for test cases that do not
-    # run the python interpreter selection task.
-    if interpreter and interpreter.version >= (3, 0 ,0):
-      self.context.log.info('Linting is currently disabled for Python 3 targets.\n '
+    python_tgts = self.context.targets(
+      lambda tgt: isinstance(tgt, (PythonTarget))
+    )
+    if not python_tgts:
+      return
+    interpreter_cache = PythonInterpreterCache(PythonSetup.global_instance(),
+                                               PythonRepos.global_instance(),
+                                               logger=self.context.log.debug)
+    tgts_by_compatibility, _ = interpreter_cache.partition_targets_by_compatibility(python_tgts)
+
+    # Check for Python 3-only constraints.
+    context_has_py3_targets = False
+    for constraint in tgts_by_compatibility.copy():
+      if '>3' in str(constraint) or '>=3' in str(constraint):
```

Comment:
This is easily foiled with spaces around operators, '>2.8,<3.7', etc. For robustness you really need to interact with constraints in `Requirement` form. Something like:
```python
requirement = Requirement.parse(constraint)
if '3' in requirement:
  # Although what about? `>=2.7` - includes 2.7 _and_ 3; ie code works with both so probably _should_ be linted under 2.7
```

Reply:
Would it be possible to intersect a "3 -> 4" range with each of them to find matches?

Reply:
Perhaps. The key point is string comparisons are not going to cut it here.